tests/test_subparser.py

```diff
@@ -0,0 +1,263 @@
+        assert config.var == 12
+        assert not config.flag
+
+    def test_args(self, capsys):
+        config = parse(self.Config, ["--flag", "Action3", "--d", "0"])
+        assert isinstance(config.mode, Action3)
+        assert config.mode.d == 0
+        assert config.flag is True
+
+        # Verify that the order matters
+        with raises(SystemExit):
+            parse(self.Config, ["Action2", "--flag"])
+        captured = capsys.readouterr()
+        assert "unrecognized arguments: --flag" in captured.err
+
+
+class TestNestedAction:
+    @dataclass
+    class Config:
+        mode: Union[Action1, Action2, Action3, Action4] = field(default_factory=lambda: Action1(0))
+        var: int = 12
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_is_not_required(self, capsys):
+        config = parse(self.Config, [])
+        assert config.mode == Action1(0)
+        assert config.var == 12
+        assert not config.flag
+
+    def test_args(self, capsys):
+        config = parse(self.Config, ["Action4", "Action1", "--a", "-3"])
+        assert isinstance(config.mode, Action4)
+        assert isinstance(config.mode.sub_action, Action1)
+        assert config.mode.sub_action.a == -3
+
+
+class TestCollision:
+    @dataclass
+    class Config:
+        action: Union[Action1, Action2]
+        a: int = 2
+        d: int = 3
+
+    def test_help(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, ["Action1", "--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert "prog Action1 [-h] --a A [--b B]" in captured.out.replace("\n", "")
+
+    def test_collision(self, capsys):
+        config = parse(self.Config, ["--a", "4", "--d", "4", "Action1", "--a", "3"])
+        assert config.action.a == 3
+        assert config.d == 4
+        assert config.a == 4
+
+    def test_other_option(self):
+        config = parse(self.Config, ["--a", "4", "--d", "4", "Action2"])
+        assert config.action.c == 42
+        assert config.d == 4
+        assert config.a == 4
+
+
+class TestPositionalAndSubparser:
+    @dataclass
+    class Config:
+        positional_1: str = field(metadata=dict(positional=True))
+        action: Union[Action1, Action2]
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+        positional_2: str = field(default="five", metadata=dict(positional=True))
+
+    def test_warn_positional_after_subparser(self, recwarn):
+        _ = parse(
+            self.Config,
+            ["positional", "Action1", "--a", "12", "--b", "20"],
+        )
+        assert len(recwarn) == 1
+
+    def test_parse_positional(self, recwarn):
+        config = parse(
+            self.Config,
+            ["positional", "Action1", "--a", "12", "--b", "20"],
+        )
+        assert config.positional_1 == "positional"
+
+
+class TestParseActionInNested:
+    @dataclass
+    class Config:
+        sub: Action4
```

Comment:
can you rename this? the nested action within Action4 is already named sub_action and this results in sub_sub_action which I find unnecessarily confusing